File: 0155-min-stack/0155-min-stack.java
class MinStack {
    
    Stack<Integer> st1;
    Stack<Integer> st2;

    public MinStack() {
        st1 = new Stack<>();
        st2 = new Stack<>();
    }
    
    public void push(int val) {
        st1.push(val);
        if(st2.empty() || val <= st2.peek()) st2.push(val);
    }
    
    public void pop() {
        if(st1.peek().equals(st2.peek())) st2.pop();
        st1.pop();  
    }
    
    public int top() {
        if(!st1.isEmpty()) return st1.peek();
        return -1; 
    }
    
    public int getMin() {
        if(!st2.isEmpty()) return st2.peek();
        return -1; 
    }
}

File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
private:
    std::vector<int> stack;
    std::vector<int> minStack;  

public:
    MinStack() {
    }
    
    void push(int val) {
        stack.push_back(val);
        
        if(minStack.empty() || val <= minStack.back()) {
            minStack.push_back(val);
        }
    }
    
    void pop() {
        if (stack.empty()) return;
        
        int topVal = stack.back();
        stack.pop_back();
        
        if (topVal == minStack.back()) {
            minStack.pop_back();
        }
    }
    
    int top() {
        if (stack.empty()) return INT_MIN; 
        return stack.back();
    }
    
    int getMin() {
        if (minStack.empty()) return INT_MIN;
        return minStack.back();  
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */